src/commonMain/kotlin/fr/acinq/lightning/message/OnionMessages.kt

```diff
@@ -11,10 +11,13 @@ import fr.acinq.lightning.utils.toByteVector
 import fr.acinq.lightning.wire.*
 
 object OnionMessages {
-    data class IntermediateNode(val nodeId: PublicKey, val outgoingChannelId: ShortChannelId? = null, val padding: ByteVector? = null, val customTlvs: Set<GenericTlv> = setOf()) {
+    data class IntermediateNode(val nodeId: PublicKey, val isPhoenix: Boolean = false, val outgoingChannelId: ShortChannelId? = null, val padding: ByteVector? = null, val customTlvs: Set<GenericTlv> = setOf()) {
```

Comment:
It isn't this `IntermediateNode` that is phoenix, it's rather the next node? Since `toTlvStream` already takes an `EncodedNodeId` that contains this information, do we really need a new field here? Would it help if we changed `val nodeId: PublicKey` to `val nodeId: EncodedNodeId`? We would probably need to slightly change the `EncodedNodeId` class to be something like:

```
sealed class EncodedNodeId {
    sealed class WithPublicKey : EncodedNodeId() {
        abstract val publicKey: PublicKey

        data class Plain(override val publicKey: PublicKey) : WithPublicKey() {
            override fun toString(): String = publicKey.toString()
        }

        data class Wallet(override val publicKey: PublicKey) : WithPublicKey() {
            override fun toString(): String = publicKey.toString()
        }
    }

    /** For compactness, nodes may be identified by the shortChannelId of one of their public channels. */
    data class ShortChannelIdDir(val isNode1: Boolean, val scid: ShortChannelId) : EncodedNodeId() {
        override fun toString(): String = if (isNode1) "<-$scid" else "$scid->"
    }
}
```

And use `val nodeId: EncodedNodeId.WithPublicKey`. It could be cleaner than boolean flags that we need to propagate everywhere?

Reply:
> It isn't this `IntermediateNode` that is phoenix, it's rather the next node?

No, it is about this node, not the next one. As you've noticed `toTlvStream` already takes this information for the next node.

> And use `val nodeId: EncodedNodeId.WithPublicKey`. It could be cleaner than boolean flags that we need to propagate everywhere?

Done, it's much nicer indeed.